python_ta/checkers/bad_loop_indexing_checker.py

```diff
@@ -0,0 +1,108 @@
+    __implements__ = IAstroidChecker
+    # name is the same as file name but without _checker part
+    name = 'bad_loop_indexing'
+    # use dashes for connecting words in message symbol
+    msgs = {'E9994': ('Bad loop indexing used in this for-loop',
+                      'bad-loop-indexing',
+                      'Used when you have an iteration variable in a for-loop '
+                      'where its only usage is to index the iterable it is iterating over'),
+            }
+
+    # this is important so that your checker is executed before others
+    priority = -1
+
+    def _check_bad_loop_indexing(self, node):
+        """Return whether the usage of the indexing variable in the for-loop is ONLY to index the iterable.
+        True if bad, False if not bad.
+        """
+        # Store the iteration variable
+        index = node.target.name
+
+        # Check if the iterable of the for-loop is of the form "range(len(lst))"
+        iterable_lst = _check_iterable_is_range(node)
+        if iterable_lst:
+            iterable = iterable_lst[0]
+        else:
+            return False
+
+        # Use node_of_class to find all Name nodes.
+        if any(_check_correct_usage(name_node, index, iterable) for name_node in node.nodes_of_class(astroid.Name)):
+            return False
+        # There is at least one 'index' Name node and they are all used badly.
+        elif any(name_node.name == index for name_node in node.nodes_of_class(astroid.Name)):
+            return True
+        # There are no 'index' Name nodes.
+        # Bad-loop-indexing checker should not catch an error since this is a case for
+        # checker W0612(unused-variable) to prevent duplicating error detection.
+        else:
+            return False
+
+    # pass in message symbol as a parameter of check_messages
+    @check_messages("bad-loop-indexing")
+    def visit_for(self, node):
+        if self._check_bad_loop_indexing(node):
+            self.add_message('bad-loop-indexing', node=node)
+
+
+# Helper functions
+def _check_iterable_is_range(node: astroid.For) -> List[str]:
+    """Return a list containing the iterable of this For node if it is of the form: range(len(iterable)).
+    Return an empty list if it is not of this form.
+    """
+    _iter = node.iter
+    # Order of operations is important
+    if isinstance(_iter, astroid.Call) and isinstance(_iter.func, astroid.Name) \
+            and _iter.func.name == 'range' and isinstance(_iter.args[0], astroid.Call) \
+            and isinstance(_iter.args[0].func, astroid.Name) and _iter.args[0].func.name == 'len':
+        return [_iter.args[0].args[0].name]
+    else:
+        return []
+
+
+def _check_correct_usage(node: astroid.Name, index: str, iterable: str) -> bool:
+    """Return whether or not this Name node 'index' was used correctly in the for loop.
+    Return True if correctly used, False if badly used or node is not 'index'.
+    """
+    # If Name node is 'index'
```

Comment:
Avoid the deep nesting here by using an [early return](http://blog.timoxley.com/post/47041269194/avoid-else-return-early)